tracer/src/Datadog.Trace/Debugger/ExceptionAutoInstrumentation/ExceptionNormalizer.cs

```diff
@@ -0,0 +1,107 @@
+        internal int NormalizeAndHashException(string exceptionString, string outerExceptionType, string? innerExceptionType)
+        {
+            if (string.IsNullOrEmpty(exceptionString))
+            {
+                throw new ArgumentException(@"Exception string cannot be null or empty", nameof(exceptionString));
+            }
+
+            var fnvHashCode = Fnv1aHash.FnvOffsetBias;
+
+            fnvHashCode = HashLine(VendoredMicrosoftCode.System.MemoryExtensions.AsSpan(outerExceptionType), fnvHashCode);
+
+            if (innerExceptionType != null)
+            {
+                fnvHashCode = HashLine(VendoredMicrosoftCode.System.MemoryExtensions.AsSpan(innerExceptionType), fnvHashCode);
+            }
+
+            var exceptionSpan = VendoredMicrosoftCode.System.MemoryExtensions.AsSpan(exceptionString);
+
+            var inSpan = VendoredMicrosoftCode.System.MemoryExtensions.AsSpan(" in ");
+
+            while (!exceptionSpan.IsEmpty)
+            {
+                var lineEndIndex = exceptionSpan.IndexOfAny('\r', '\n');
+                VendoredMicrosoftCode.System.ReadOnlySpan<char> line;
+
+                if (lineEndIndex >= 0)
+                {
+                    line = exceptionSpan.Slice(0, lineEndIndex);
+                    exceptionSpan = exceptionSpan.Slice(lineEndIndex + 1);
+                    if (!exceptionSpan.IsEmpty && exceptionSpan[0] == '\n')
+                    {
+                        exceptionSpan = exceptionSpan.Slice(1);
+                    }
+                }
+                else
+                {
+                    line = exceptionSpan;
+                    exceptionSpan = default;
+                }
+
+                if (IsStackTraceLine(line) && !IsLambdaFunctionLine(line))
+                {
+                    var index = VendoredMicrosoftCode.System.MemoryExtensions.IndexOf(line, inSpan);
+                    fnvHashCode = HashLine(index > 0 ? line.Slice(0, index) : line, fnvHashCode);
+                }
+            }
+
+            return fnvHashCode;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool IsStackTraceLine(VendoredMicrosoftCode.System.ReadOnlySpan<char> line)
+        {
+            int i = 0;
+
+            while (i < line.Length && char.IsWhiteSpace(line[i]))
+            {
+                i++;
+            }
+
+            return line.Length - i >= 3 && line[i] == 'a' && line[i + 1] == 't' && line[i + 2] == ' ';
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool IsLambdaFunctionLine(VendoredMicrosoftCode.System.ReadOnlySpan<char> line)
+        {
+            var lambdaSpan = VendoredMicrosoftCode.System.MemoryExtensions.AsSpan("lambda_");
+            return VendoredMicrosoftCode.System.MemoryExtensions.Contains(line, lambdaSpan, StringComparison.Ordinal);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```

Comment:
The `AggressiveInlinig` might be ignored here because the the method is virtual (late binding). I see that you made it virtual for testing, is it really necessary?